src/library/export/engineprimeexportjob.cpp

```diff
@@ -0,0 +1,532 @@
+#include "library/trackset/crate/crate.h"
+#include "track/track.h"
+#include "util/optional.h"
+#include "waveform/waveformfactory.h"
+
+namespace el = djinterop::enginelibrary;
+
+namespace mixxx {
+
+namespace {
+
+const std::string kMixxxRootCrateName = "Mixxx";
+
+const QStringList kSupportedFileTypes = {"mp3", "flac", "ogg"};
+
+std::optional<djinterop::musical_key> toDjinteropKey(
+        track::io::key::ChromaticKey key) {
+    static const std::array<std::optional<djinterop::musical_key>, 25> keyMap{{
+            std::nullopt,                          // INVALID = 0,
+            djinterop::musical_key::c_major,       // C_MAJOR = 1,
+            djinterop::musical_key::d_flat_major,  // D_FLAT_MAJOR = 2,
+            djinterop::musical_key::d_major,       // D_MAJOR = 3,
+            djinterop::musical_key::e_flat_major,  // E_FLAT_MAJOR = 4,
+            djinterop::musical_key::e_major,       // E_MAJOR = 5,
+            djinterop::musical_key::f_major,       // F_MAJOR = 6,
+            djinterop::musical_key::f_sharp_major, // F_SHARP_MAJOR = 7,
+            djinterop::musical_key::g_major,       // G_MAJOR = 8,
+            djinterop::musical_key::a_flat_major,  // A_FLAT_MAJOR = 9,
+            djinterop::musical_key::a_major,       // A_MAJOR = 10,
+            djinterop::musical_key::b_flat_major,  // B_FLAT_MAJOR = 11,
+            djinterop::musical_key::b_major,       // B_MAJOR = 12,
+            djinterop::musical_key::c_minor,       // C_MINOR = 13,
+            djinterop::musical_key::d_flat_minor,  // C_SHARP_MINOR = 14,
+            djinterop::musical_key::d_minor,       // D_MINOR = 15,
+            djinterop::musical_key::e_flat_minor,  // E_FLAT_MINOR = 16,
+            djinterop::musical_key::e_minor,       // E_MINOR = 17,
+            djinterop::musical_key::f_minor,       // F_MINOR = 18,
+            djinterop::musical_key::f_sharp_minor, // F_SHARP_MINOR = 19,
+            djinterop::musical_key::g_minor,       // G_MINOR = 20,
+            djinterop::musical_key::a_flat_minor,  // G_SHARP_MINOR = 21,
+            djinterop::musical_key::a_minor,       // A_MINOR = 22,
+            djinterop::musical_key::b_flat_minor,  // B_FLAT_MINOR = 23,
+            djinterop::musical_key::b_minor,       // B_MINOR = 24
+    }};
+
+    return keyMap[key];
+}
+
+std::string exportFile(const EnginePrimeExportRequest& request,
+        TrackPointer pTrack) {
+    if (!request.engineLibraryDbDir.exists()) {
+        auto msg = QString(
+                "Engine Library DB directory %1 has been removed from disk!")
+                           .arg(request.engineLibraryDbDir.absolutePath());
+        throw std::runtime_error{msg.toStdString()};
+    } else if (!request.musicFilesDir.exists()) {
+        auto msg = QString(
+                "Music file export directory %1 has been removed from disk!")
+                           .arg(request.musicFilesDir.absolutePath());
+        throw std::runtime_error{msg.toStdString()};
```

Comment:
We don't use exceptions. If your library uses exceptions then enclose the code in a try/catch handler somewhere. I haven't seen one?

Reply:
Exceptions were previously propagating out - I've now added handlers in the `run()` method, along with a signal and dialog to inform the user if an error occurs.